src/System.Buffers/src/System/Buffers/BufferData.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace System.Buffers
+{
+    public struct BufferData
+    {
+        internal byte[] _data;
+        internal bool _allocd;
```

Comment:
I do not think you need to track _allocd. You are using this to not accept back to the pool buffers that did not come from the pool, but I think as long as there is a space in the pool and the buffer is not larger than _maxBufferSize you could just take it. They are all just byte arrays.